### PR TITLE
Project creators receive email notification when their project is approved or rejected

### LINKED ISSUE #65: Project members receive an email when their project is approved or rejected by an admin
While working on #13 it became pretty clear that getting email stuff working should be it's own PR since it is pretty meaty already.

We'll probably want to add the line `Then the project members receive a project approval email` to the `Admin approves a project` Scenario.

### PR DESCRIPTION
**Pull request by zspencer**
Resolves #65

### CODE REVIEW

File: app/models/project.rb [hunk 1]
```diff
@@ -8,25 +8,40 @@ class Project < ApplicationRecord
...
   def rejected?
     status == "rejected"
   end
 
-  def pending?
-    status == "pending"
+  def apply_status_change(status_change, status:)
+    status_change.project = self
+    transaction { status_change.save && update(status: status) }
   end
```
Comment (zspencer) [inline app/models/project.rb]: This should probably be private.

File: app/mailers/project_status_change_mailer.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# Sends emails to notify project team members when their project status has changed
+# @see http://localhost:3000/rails/mailers/project_status_change_mailer/
+class ProjectStatusChangeMailer < ApplicationMailer
+  def approved(member:, approval:)
+    mail(to: member.email) do |format|
+      format.html { render locals: { approval: approval, member: member } }
+      format.text { render locals: { approval: approval, member: member } }
```
Comment (zspencer) [inline app/mailers/project_status_change_mailer.rb]: This is because I didn't want to define `@vars` all over the place; but I'm fine to do that since it is more conventional.

File: config/routes.rb [hunk 3]
```diff
@@ -3,13 +3,13 @@
...
+  resources :projects, only: %i[new create] do
+    resources :approvals, only: %i[new create], controller: "project_rejections"
+    resources :rejections, only: %i[new create], controller: "project_approvals"
+  end
   resources :my_projects, only: %i[index]
 
-  resources :pending_projects, only: [:index] do
-    resources :project_approvals, only: %i[new create]
-    resources :project_rejections, only: %i[new create]
-  end
```
Comment (zspencer) [inline config/routes.rb]: Pulled these out of the pending_projects namespace because it was getting a little bit absurd with the `pending_project_project_...` Happy to change it back but it seemed like a reasonable move at the time.

File: features/support/pages/app.rb [hunk 4]
```diff
@@ -111,4 +111,17 @@ def public_project?(project: nil, title: nil)
     current_page.project?(title: title)
   end
   alias has_public_project? public_project?
+
+  def emails_to(address)
+    ActionMailer::Base.deliveries.select do |email|
+      email.to.include?(address)
+    end.map(&TestEmail.method(:new))
```
Comment (zspencer) [inline features/support/pages/app.rb]: We may want to pull the mapping to the wrapper to a higher level, but this seemed OK for now.

Comment (zspencer) [reply features/support/pages/app.rb]: Also @bhaibel or @jtu0 - do you know of a gem that gives us this kind of stuff for free without infecting our project with a billion step definitions?

File: features/support/pages/pending_projects_page.rb [hunk 5]
```diff
@@ -15,7 +15,7 @@ def approve(project:, reason: "I just like it, OK?")
     project_status_change_form.submit(reason: reason)
   end
 
-  def reject(project:, reason: "I don't like it")
+  def reject(project:, reason: "I dislike it")
```
Comment (zspencer) [inline features/support/pages/pending_projects_page.rb]: APOSTROPHES ARE TURNED INTO `&#rsquo;` AND IT WAS CAUSING MY TEST TO FAIL AND I DIDN'T KNOW WHY

File: .env.test [hunk 6]
```diff
@@ -0,0 +1 @@
+export EMAIL_DEFAULT_FROM='no-reply@nourish.example.com'
```
Comment (zspencer) [inline .env.test]: You'll want to `source` this before running cucumber; or we could bring in dotenv for cucumber and rspec. Thoughts @bhaibel and @jtu0 ?

File: app/controllers/project_approvals_controller.rb [hunk 7]
```diff
@@ -1,25 +1,21 @@
 # Allows a project to be approved
 class ProjectApprovalsController < ModerationController
-  expose(:project_approval, scope: -> { pending_project.project_status_changes },
-                            model: :project_status_change)
-  expose(:pending_project, scope: -> { Project.pending }, model: :project)
+  expose(:approval, scope: -> { project.project_status_changes },
```
Comment (zspencer) [inline app/controllers/project_approvals_controller.rb]: This rename should have been done in a different PR but I got tripped up in it due to weird naming issues as I was making the feature work. Can either do it separately or back it out now that I know how to fix the weird naming issues.

File: app/controllers/project_approvals_controller.rb [hunk 8]
```diff
@@ -1,25 +1,21 @@
...
 
   def new; end
 
   def create
-    ProjectStatusChange.transaction do
-      pending_project.approve if project_approval.save
-    end
-
-    if project_approval.persisted?
-      flash[:notice] = t("approving_project.success_notification", project_title: pending_project.title)
```
Comment (zspencer) [inline app/controllers/project_approvals_controller.rb]: I pulled this up into the `Project` class because the create method got to the point where it was far too large; and because the `Project` is the root so it seemed like the right place to pull it instead of sprouting a branch or what not.

File: app/controllers/project_approvals_controller.rb [hunk 9]
```diff
@@ -1,25 +1,21 @@
...
+    if project.approve(approval: approval)
+      flash[:notice] = t("approving_project.success_notification", project_title: project.title)
```
Comment (zspencer) [inline app/controllers/project_approvals_controller.rb]: I kind  of want these `i18n` lookup names to match the controllers they're being used in. Thoughts?

File: app/models/project.rb [hunk 10]
```diff
@@ -8,25 +8,40 @@ class Project < ApplicationRecord
...
-    update(status: "approved")
-  end
+    return false unless apply_status_change(approval, status: "approved")
 
-  def reject
-    return true if rejected?
-    update(status: "rejected")
+    members.each do |member|
+      ProjectStatusChangeMailer.approved(member: member, approval: approval).deliver
+    end
```
Comment (zspencer) [inline app/models/project.rb]: It would be possible to move this into the `apply_status_change` method, but I wanted to separate the data mutation from the sending of the emails. I am also considering pulling the enqueuing of the emails to background job once we have background jobs setup.

File: config/application.rb [hunk 11]
```diff
@@ -27,5 +27,7 @@ class Application < Rails::Application
 
     # Don't generate system test files.
     config.generators.system_tests = nil
+
+    config.action_mailer.preview_path = Rails.root.join("spec", "mailers", "previews")
```
Comment (zspencer) [inline config/application.rb]: This may be unnecessary, I will double check and remove if it isn't needed.

File: .env.test (hunk 6 above)
Comment (zspencer) [reply .env.test]: I brought in .env to load .env.test in rspec and cucumber.

File: config/locales/en.yml [hunk 12]
```diff
@@ -38,3 +41,14 @@ en:
     connection_failed: "We were unable to connect the stripe account. Try again?"
     connection_destroyed: "The connection between nourish.party and stripe account named %{stripe_display_name} was removed."
     connection_destroy_failed: "The connection between nourish.party and the stripe account named %{stripe_display_name} was unable to be removed. Try again?"
+  project_status_change_mailer:
+    approved:
+      subject: "Your project has been approved to receive support on Nourish.Party"
+      preamble: "We're happy to announce that your project has been approved to receive support on Nourish.Party. Staff passed along the following reason:"
+      conclusion: "We're looking forward to celebrating your project's success and helping you achieve personal sustainability as a creative!"
```
Comment (jtu0) [inline config/locales/en.yml]: Minor:
Creators might not be aiming for personal sustainability.  Change to:
"We're looking forward to celebrating your project's success and helping you achieve your funding goals!"

File: features/step_definitions/project_steps.rb [hunk 13]
```diff
@@ -43,12 +43,14 @@
   end
 end
 
-When("the project is approved") do
-  app.project_under_test.approve
-end
-
-When("the project is rejected") do
-  app.project_under_test.reject
+When(/the project is (approved|rejected)/) do |action|
```
Comment (jtu0) [inline features/step_definitions/project_steps.rb]: Would changing this to
`When(/the project receives (approval|rejection)/) do |factory_name|`
allow you to skip the case and jump straight to `build(factor_name.to_sym)`?

File: features/support/helper.rb [hunk 14]
```diff
@@ -1,3 +1,5 @@
+require "dotenv"
+Dotenv.load(".env.test")
```
Comment (jtu0) [inline features/support/helper.rb]: :+1:

File: features/step_definitions/project_steps.rb (hunk 13 above)
Comment (zspencer) [reply features/step_definitions/project_steps.rb]: yep!

File: config/locales/en.yml (hunk 12 above)
Comment (zspencer) [reply config/locales/en.yml]: done.

File: app/controllers/project_approvals_controller.rb [hunk 15]
```diff
@@ -1,25 +1,21 @@
...
   def new; end
 
   def create
-    ProjectStatusChange.transaction do
-      pending_project.approve if project_approval.save
-    end
-
-    if project_approval.persisted?
-      flash[:notice] = t("approving_project.success_notification", project_title: pending_project.title)
+    if project.apply_status_change(approval)
```
Comment (bhaibel) [inline app/controllers/project_approvals_controller.rb]: I like this interface. Definitely much nicer than the factory.

File: app/mailers/project_status_change_mailer.rb [hunk 16]
```diff
@@ -0,0 +1,22 @@
+# Sends emails to notify project team members when their project status has changed
+# @see http://localhost:3000/rails/mailers/project_status_change_mailer/
+class ProjectStatusChangeMailer < ApplicationMailer
```
Comment (bhaibel) [inline app/mailers/project_status_change_mailer.rb]: Not in love with having the same mailer for both kinds of status changes. I'd prefer separating this back out; I'm concerned having the abstraction present in the mailers at first is going to obscure stuff (particularly around appropriate messaging structure.)

Nonblocking.

File: config/routes.rb (hunk 3 above)
Comment (bhaibel) [reply config/routes.rb]: maybe `resources :project_approvals, as: :approvals` ?

File: features/step_definitions/project_steps.rb [hunk 17]
```diff
@@ -43,12 +43,8 @@
   end
 end
 
-When("the project is approved") do
-  app.project_under_test.approve
-end
-
-When("the project is rejected") do
-  app.project_under_test.reject
+When(/the project receives (approval|rejection)/) do |status_change_factory|
```
Comment (bhaibel) [inline features/step_definitions/project_steps.rb]: This feels like DRY for DRY's sake.

Comment (zspencer) [reply features/step_definitions/project_steps.rb]: It's DRY because I had to change it several times.

File: app/mailers/project_status_change_mailer.rb (hunk 16 above)
Comment (zspencer) [reply app/mailers/project_status_change_mailer.rb]: I'd be fine pulling it back out into separate mailers. I prefer one to one mappings of mailers to actual emails sent as well; and hiding the abstraction at the locales level feels really weird to me.

I'd rather have `ProjectApprovedMailer` and `ProjectRejectedMailer` and switch between them using a boolean.
